### PR TITLE
feat(chip-set): make it possible to set `input` type of `search` or `text`

### PR DESCRIPTION
**Pull request by LucyChyzhova**
fix: #1718

## Review:
- [ ] Commits are [atomic](https://seesparkbox.com/foundry/atomic_commits_with_git)
- [ ] Commits have the correct *type* for the changes made
- [ ] Commits with *breaking changes* are marked as such

### Browsers tested:
(Check any that applies, it's ok to leave boxes unchecked if testing something didn't seem relevant.)

Windows:
- [ ] Chrome
- [ ] Edge
- [ ] Firefox

Linux:
- [ ] Chrome
- [ ] Firefox

macOS:
- [ ] Chrome
- [ ] Firefox
- [ ] Safari

Mobile:
- [ ] Chrome on Android
- [ ] iOS

### CODE REVIEW

File: src/components/chip-set/chip-set.tsx
Comment (Kiarokh) [inline src/components/chip-set/chip-set.tsx:397]: Image from iOS

## ⚠️ head up

We shouldn't merge this yet, until we know how to get rid of the extra 🔍 icon that iOS adds.

Comment (Kiarokh) [reply src/components/chip-set/chip-set.tsx:397]: https://stackoverflow.com/questions/23296072/html5-search-input-magnifying-glass-override/23296152

Comment (LucyChyzhova) [reply src/components/chip-set/chip-set.tsx:397]: The last changes fixed an extra X button and the background colour 
 Image from iOS (1)
Image from iOS

File: src/components/chip-set/chip-set.tsx
Comment (LucyChyzhova) [reply src/components/chip-set/chip-set.tsx:91]: Sounds good! I will create a new example and put this long description there.

File: src/components/chip-set/chip-set.tsx
Comment (adrianschmidt) [inline src/components/chip-set/chip-set.tsx:396]: Sorry for being off topic to the PR, but are we still setting tab-index to 1 here? tab-index should only ever be set to `-1` or `0`. I pointed this out before, but can't remember if there was some reason we couldn't just set it to `0`? @Kiarokh

Comment (Kiarokh) [reply src/components/chip-set/chip-set.tsx:385]: I don't know really. But I made this issue now so we don't forget to test this: https://github.com/Lundalogik/lime-elements/issues/1741

File: src/components/chip-set/examples/chip-set-input-type-search.tsx
Comment (adrianschmidt) [inline src/components/chip-set/examples/chip-set-input-type-search.tsx:1]: I squashed the commit with these examples into the feature commit, as I don't think they need to be separate, since it's docs for the feature you added.

I also added a fixup where I reworded some of the descriptions a little.

I've only tested this quickly on an iPhone on Browserstack, but didn't get any autocorrection there regardless of input type… 🤔

I'd like to test a little bit more, but I need to log off for today now.

File: src/components/chip-set/chip-set.tsx
Comment (adrianschmidt) [reply src/components/chip-set/chip-set.tsx:382]: So, what's the status here @Kiarokh? OK to merge after review, or not?

File: src/components/chip-set/examples/chip-set-input-type-search.tsx
Comment (adrianschmidt) [reply src/components/chip-set/examples/chip-set-input-type-search.tsx:1]: On Browserstack, I'm not getting autocorrection on either iOS or Android, regardless of whether I test the `text` or the `search` example.

I am getting the magnifying glass icon on the enter key on Android though. For the picker I guess that's good. For the chip-set example it seems quite misplaced though…

Using `search`:
A screenshot of the limel-example-chip-set-input-type-search component with Android's software keyboard open, showing a magnifying glass icon on the Enter key

Using `text`:
A screenshot of the limel-example-chip-set-input-type-text component with Android's software keyboard open, showing an arrow-pointing-right icon on the Enter key

I do think we should allow setting the type regardless, so we can use `search` for the picker, but I think we might also want to allow some more granular control, like that described [here](https://davidwalsh.name/disable-autocorrect). That would be something we'd first enable in limel-input though…

Comment (adrianschmidt) [reply src/components/chip-set/examples/chip-set-input-type-search.tsx:1]: Ah, we don't use limel-input in the chips-set, right? So, no, we don't need to implement anything on limel-input before implementing it here 🙂

Comment (LucyChyzhova) [reply src/components/chip-set/examples/chip-set-input-type-search.tsx:1]: > we don't use limel-input in the chips-set, right?

No, we don't use limel-input here.

File: src/components/chip-set/examples/chip-set-input-type-text.tsx
Comment (adrianschmidt) [inline src/components/chip-set/examples/chip-set-input-type-text.tsx:25]: Bias declaration: I don't use automatic autocorrection, because I prefer to just get a red underline for potential typos, rather than having my phone change stuff automatically, because I just can't stand when it "autocorrects" something away from what I actually meant to something else. Given that, I have a hard time imagining what a user who likes autocorrection might prefer here. But:

I assume that, if you have autocorrection enabled, if you actually meant what you wrote, you can go back and change it again (I guess you can, right? It won't then autocorrect _again_?). If so, what about when it's used in the chip-set? If you write something that gets incorrectly "autocorrected", remove the incorrect chip, and then write the same thing again, will your phone understand not to change it this time, or will it "autocorrect" it again, since you did "other stuff" in between?

If it would "autocorrect" again, having autocorrection enabled on a chip-set would make it very hard to write anything the autocorrection feature doesn't "allow"…

I guess you could write your word, press space to have it autocorrect, then backspace and correct what you wrote, and only _then_ press enter. But how many users are actually going to figure that out?

As I said, I'm clearly biased, and I'm not even going to be subjected to any of this, since I have autocorrection disabled, but I feel like having autocorrection on a chip-set is exactly the kind of thing that could make users hate our app… 😞

File: src/components/chip-set/examples/chip-set-input-type-search.tsx
Comment (Kiarokh) [reply src/components/chip-set/examples/chip-set-input-type-search.tsx:1]: sorry for the delayed response. Yes the granular controls is a great idea. Actually both in here and in limel-input. :) But setting type to search sounds also like these granular controls. no?

> image

interesting effect here. I didn't know it works like that in Android.

> On Browserstack, I'm not getting autocorrection on either iOS or Android,

I think they have turned these features off on their keyboard settings. Probably they were problematic, given the service is used world-wide.

File: src/components/chip-set/chip-set.tsx
Comment (Kiarokh) [reply src/components/chip-set/chip-set.tsx:382]: I think it's alright.

File: src/components/chip-set/examples/chip-set-input-type-text.tsx
Comment (Kiarokh) [reply src/components/chip-set/examples/chip-set-input-type-text.tsx:25]: > you can go back and change it again (I guess you can, right? It won't then autocorrect again?) […] 

It depends on the device, and timing of when you correct it, where you are typing, and some other factors that I don't really figure out. But it will in many cases. 

>  […] what about chipset […]

I think this also depends on device, timing, etc. On iOS when I try, the browser stops auto-correcting (for a while[?] while the `focus` is still in that field[?]) when I remove the wrong chip or word and try again. But the primary problem here is that user must realize first that what's in the "chip" is not what they wrote. For my dyslexic brain for instance, this may take a very long time sometimes. 

Note that chips don't show that red underline that indicates spelling errors.

> But how many users are actually going to figure that out?

If you realize that this autocorrection happens, you can find a way to hack it. But first you should consciously realize it.

File: src/components/chip-set/examples/chip-set-input-type-search.tsx
Comment (adrianschmidt) [reply src/components/chip-set/examples/chip-set-input-type-search.tsx:1]: > sorry for the delayed response. Yes the granular controls is a great idea. Actually both in here and in limel-input. :) But setting type to search sounds also like these granular controls. no?

> I do think we should allow setting the type regardless, so we can use `search` for the picker

😄

> I think they have turned these features off on their keyboard settings. Probably they were problematic, given the service is used world-wide.

Makes sense I guess… it's just a pity if it can't be turned on and off so it can be tested 😞 But I haven't looked, maybe there _is_ a setting for it 🤷‍♂️

File: src/components/chip-set/chip-set.tsx
Comment (adrianschmidt) [inline src/components/chip-set/chip-set.tsx:95]:
```suggestion
     * For chip-sets of type `input`. Value to use for the `type` attribute on the
     * input field inside the chip-set.
```

File: src/components/chip-set/examples/chip-set-input-type-text.tsx
Comment (adrianschmidt) [reply src/components/chip-set/examples/chip-set-input-type-text.tsx:25]: I agree with your assessment above. It's unlikely the user will notice any autocorrection will happen, and it's even less likely they will figure out how to get around it, in case the autocorrect didn't do what they wanted. So I think we should just disable autocorrect for all limel-chip-sets, regardless of whether the input type is text or search…

I think we can do that either in this PR or separately, but if we want to do that, it seems quite unnecessary to now add a long explanation of why enabling autocorrection is good in some cases. That's the only reason I'm bringing this up here, rather than just merging 🙂

File: src/components/chip-set/examples/chip-set-input-type-search.tsx
Comment (Kiarokh) [reply src/components/chip-set/examples/chip-set-input-type-search.tsx:1]: yes. When we tested on a real iOS device, the behavior changed as follows:

- The auto-correction would not automatically kick in before pressing space or enter. 
- the keyboard's predictive and spelling suggestions would still show word suggestions on keyboard's top bar. So user would notice that they could pick another word, or some spelling might be wrong. Like this image 👇 
image

I don't know how Android behaves though

File: src/components/chip-set/examples/chip-set-input-type-text.tsx
Comment (Kiarokh) [reply src/components/chip-set/examples/chip-set-input-type-text.tsx:25]: yes makes sense. If we want to introduce `auto-correction= off` then we should not explain these stuff for input type. And at least I don't know a if we will have any more legit reasons to keep input type (we're overwriting the browser styles anyways). Or do you think we should keep both as props?

Comment (adrianschmidt) [reply src/components/chip-set/examples/chip-set-input-type-text.tsx:25]: We could keep both, but if the main point is to disable autocorrection, it would be a smaller change to just add that in limel-chip-set, without changing the interface at all. Then we can always add the inputType option if we feel an actual need for it.

Comment (Kiarokh) [reply src/components/chip-set/examples/chip-set-input-type-text.tsx:25]: sounds good to me

Comment (Kiarokh) [reply src/components/chip-set/examples/chip-set-input-type-text.tsx:25]: what do you think @LucyChyzhova ?

Comment (Kiarokh) [reply src/components/chip-set/examples/chip-set-input-type-text.tsx:25]: Oh actually when I read [this](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/search#search_form_labels_and_accessibility) again, I change my mind. Autocorrection is one of the things we want to remove, when the user action is about searching. But that doesn't solve the accessibility problems.

So in this search field 
image
a screen reader wouldn't give enough information to users. Specially because our label ("Begins with") is also confusing. It becomes more clear for a sighted user, when we show the magnifying glass icon there. But not-sighted users won't see that. They would need more clues to realize they can actually perform a search.

Comment (adrianschmidt) [reply src/components/chip-set/examples/chip-set-input-type-text.tsx:25]: The way you describe that autocorrection works for limel-chip-set, it _really_ doesn't sound to me like we would want autocorrection there, since it would be pretty much impossible for a user to write something that their phone thinks it ought to autocorrect, in case the user doesn't actually want it "corrected".

But I'll leave that up to you, since you are our interaction designer, after all 😊

**Review by Kiarokh — APPROVED** [review]
I think it's good to go with this solution, and in another PR add new features to `limel-input` (or other plain _non-MDC_ input fields which we might have in some components), to enable things like:

````html
<input 
	autocomplete="off" 
	autocorrect="off" 
	autocapitalize="off" 
	spellcheck="false" 
/>
````
Maybe we get these when we upgrade MDC, or maybe they're just there already for us to activate as new features.

And I'm not sure how big of a demand is out there for such features. I don't think at least us in the CRM would need them any time soon? Or maybe we do, but never have thought about them. Maybe it's good for instance to activate `autocorrect="off"` for the company name input field. 🤷‍♂️

**Pull request merged**
